File: old/services/template.py
"""
Template Service - manages template operations on Proxmox host
"""
import logging
from typing import Optional
from libs.config import LabConfig
from .lxc import LXCService
logger = logging.getLogger(__name__)

class TemplateService:
    """Service for managing templates on Proxmox host"""
    def __init__(self, lxc_service: LXCService):
        """
        Initialize Template service
        Args:
            lxc_service: LXC service instance with SSH connection
        """
        self.lxc = lxc_service

    def get_base_template(self, cfg: LabConfig) -> Optional[str]:
        """
        Get base Ubuntu template, download if needed
        Args:
            cfg: Lab configuration
        Returns:
            Template name or None if download failed
        """
        templates = cfg.template_config.base
        template_dir = cfg.proxmox_template_dir
        for template in templates:
            check_cmd = f"test -f {template_dir}/{template} && echo exists || echo missing"
            check_result, _ = self.lxc.execute(check_cmd)
            if check_result and "exists" in check_result:
                return template
        # Download last template in list
        template_to_download = templates[-1]
        logger.info("Base template not found. Downloading %s...", template_to_download)
        # Run pveam download with live output
        download_cmd = f"pveam download local {template_to_download}"
        logger.info("Running: %s", download_cmd)
        output, exit_code = self.lxc.execute(download_cmd, timeout=300)
        # Verify download completed
        verify_cmd = f"test -f {template_dir}/{template_to_download} && echo exists || echo missing"
        verify_result, _ = self.lxc.execute(verify_cmd)
        if not verify_result or "exists" not in verify_result:
            error_msg = f"Template {template_to_download} download failed"
            if exit_code is not None and exit_code != 0:
                error_msg += f" (exit code: {exit_code})"
            if output:
                error_msg += f". Output: {output}"
            logger.error(error_msg)
            raise RuntimeError(error_msg)
        logger.info("Template %s downloaded successfully", template_to_download)
        return template_to_download

    def get_template_path(self, template_name: Optional[str], cfg: LabConfig) -> str:
        """
        Get path to template file by template name
        Args:
            template_name: Template name or None for base template
            cfg: Lab configuration
        Returns:
            Full path to template file
        Raises:
            RuntimeError: If template cannot be found or downloaded
        """
        template_dir = cfg.proxmox_template_dir
        # If template_name is None, use base template directly
        if template_name is None:
            base_template = self.get_base_template(cfg)
            if not base_template:
                raise RuntimeError("Failed to get base template")
            return f"{template_dir}/{base_template}"
        # Find template config
        template_cfg = None
        for tmpl in cfg.templates:
            if tmpl.name == template_name:
                template_cfg = tmpl
                break
        if not template_cfg:
            # Fallback to base template
            base_template = self.get_base_template(cfg)
            if not base_template:
                raise RuntimeError("Failed to get base template")
            return f"{template_dir}/{base_template}"
        # Find template file by name - search for files matching template name
        template_name_pattern = f"{template_cfg.name}*.tar.zst"
        find_cmd = f"ls -t {template_dir}/{template_name_pattern} 2>/dev/null | head -1 | xargs basename 2>/dev/null"
        template_file, _ = self.lxc.execute(find_cmd)
        if template_file:
            return f"{template_dir}/{template_file.strip()}"
        # Fallback to base template
        base_template = self.get_base_template(cfg)
        if not base_template:
            raise RuntimeError("Failed to get base template")
        return f"{template_dir}/{base_template}"

    def validate_template(self, template_path: str) -> bool:
        """
        Validate that template file exists and is readable
        Args:
            template_path: Full path to template file
        Returns:
            True if template is valid, False otherwise
        """
        validate_cmd = f"test -f {template_path} && test -r {template_path} && echo 'valid' || echo 'invalid'"
        result, _ = self.lxc.execute(validate_cmd)
        return result is not None and "valid" in result

File: src/services/template.go
package services

import (
	"enva/libs"
	"fmt"
	"strings"
)

// TemplateService manages template operations on LXC host
type TemplateService struct {
	lxc libs.LXCServiceInterface
}

// NewTemplateService creates a new Template service
func NewTemplateService(lxc libs.LXCServiceInterface) *TemplateService {
	return &TemplateService{lxc: lxc}
}

// GetBaseTemplate gets base Ubuntu template, download if needed
func (t *TemplateService) GetBaseTemplate(cfg *libs.LabConfig) *string {
	templates := cfg.TemplateConfig.Base
	templateDir := cfg.LXCTemplateDir()

	for _, template := range templates {
		checkCmd := fmt.Sprintf("test -f %s/%s && echo exists || echo missing", templateDir, template)
		checkResult, _ := t.lxc.Execute(checkCmd, nil)
		if strings.Contains(checkResult, "exists") {
			return &template
		}
	}

	// Download last template in list
	if len(templates) == 0 {
		return nil
	}
	templateToDownload := templates[len(templates)-1]
	libs.GetLogger("template").Printf("Base template not found. Downloading %s...", templateToDownload)

	// Run pveam download
	downloadCmd := fmt.Sprintf("pveam download local %s", templateToDownload)
	libs.GetLogger("template").Debug("Running: %s", downloadCmd)
	timeout := 300
	output, exitCode := t.lxc.Execute(downloadCmd, &timeout)

	// Verify download completed
	verifyCmd := fmt.Sprintf("test -f %s/%s && echo exists || echo missing", templateDir, templateToDownload)
	verifyResult, _ := t.lxc.Execute(verifyCmd, nil)
	if !strings.Contains(verifyResult, "exists") {
		errorMsg := fmt.Sprintf("Template %s download failed", templateToDownload)
		if exitCode != nil && *exitCode != 0 {
			errorMsg += fmt.Sprintf(" (exit code: %d)", *exitCode)
		}
		if output != "" {
			errorMsg += fmt.Sprintf(". Output: %s", output)
		}
		libs.GetLogger("template").Printf(errorMsg)
		return nil
	}
	libs.GetLogger("template").Printf("Template %s downloaded successfully", templateToDownload)
	return &templateToDownload
}

// GetTemplatePath gets template path for given template name
func (t *TemplateService) GetTemplatePath(templateName *string, cfg *libs.LabConfig) string {
	templateDir := cfg.LXCTemplateDir()
	if templateName == nil {
		// Use base template
		baseTemplate := t.GetBaseTemplate(cfg)
		if baseTemplate == nil {
			return ""
		}
		return fmt.Sprintf("%s/%s", templateDir, *baseTemplate)
	}
	// Check if it's a template file name or a template name
	if strings.Contains(*templateName, ".tar") {
		// It's a template file
		return fmt.Sprintf("%s/%s", templateDir, *templateName)
	}
	// It's a template name, find the file matching pattern (matching Python logic)
	// Find template config
	var templateCfg *libs.TemplateConfig
	for _, tmpl := range cfg.Templates {
		if tmpl.Name == *templateName {
			templateCfg = &tmpl
			break
		}
	}
	if templateCfg == nil {
		// Fallback to base template
		baseTemplate := t.GetBaseTemplate(cfg)
		if baseTemplate == nil {
			return ""
		}
		return fmt.Sprintf("%s/%s", templateDir, *baseTemplate)
	}
	// Find template file by name - search for files matching template name pattern (matching Python: ls -t {template_name}*.tar.zst)
	templateNamePattern := fmt.Sprintf("%s*.tar.zst", templateCfg.Name)
	findCmd := fmt.Sprintf("ls -t %s/%s 2>/dev/null | head -1 | xargs basename 2>/dev/null", templateDir, templateNamePattern)
	templateFile, _ := t.lxc.Execute(findCmd, nil)
	if templateFile != "" && strings.TrimSpace(templateFile) != "" {
		return fmt.Sprintf("%s/%s", templateDir, strings.TrimSpace(templateFile))
	}
	// Fallback to base template
	baseTemplate := t.GetBaseTemplate(cfg)
	if baseTemplate == nil {
		return ""
	}
	return fmt.Sprintf("%s/%s", templateDir, *baseTemplate)
}

// ValidateTemplate validates template file exists and is readable
func (t *TemplateService) ValidateTemplate(templatePath string) bool {
	checkCmd := fmt.Sprintf("test -f %s && test -r %s && echo 'valid' || echo 'invalid'", templatePath, templatePath)
	result, _ := t.lxc.Execute(checkCmd, nil)
	return strings.Contains(result, "valid")
}
